app/src/main/java/pl/droidsonroids/droidsmap/MapActivity.kt

```diff
@@ -1,31 +1,68 @@
 package pl.droidsonroids.droidsmap
 
+import android.Manifest
+import android.content.pm.PackageManager
 import android.os.Bundle
+import android.support.v4.app.ActivityCompat
+import android.support.v4.content.ContextCompat
 import android.support.v7.app.AppCompatActivity
+import android.widget.Toast
 import pl.droidsonroids.droidsmap.feature.office.ui.OfficeUiFeatureView
 import pl.droidsonroids.droidsmap.feature.room.ui.RoomUiFeatureView
 
+private const val LOCATION_REQUEST_CODE = 1
+
 class MapActivity : AppCompatActivity() {
 
+    private lateinit var flowManager: FlowManager
     private lateinit var officeFeature: OfficeUiFeatureView
     private lateinit var roomFeature: RoomUiFeatureView
 
+    val appTerminateCallback = object : TerminationCallback {
+        override fun onAppTerminate() {
+            this@MapActivity.finish()
+        }
+    }
+
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
         setContentView(R.layout.activity_map)
-
         officeFeature = OfficeUiFeatureView(this)
-        officeFeature.requestOffice()
         roomFeature = RoomUiFeatureView(this)
+        flowManager = FlowManager(officeFeature, roomFeature, appTerminateCallback)
+
+        checkLocationPermission()
     }
 
-    override fun onBackPressed() {
-        officeFeature.onBackButtonPressed()
-        roomFeature.onBackButtonPressed()
+    private fun checkLocationPermission() {
+        if (ContextCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED) {
+            officeFeature.onLocationPermissionGranted()
+        } else {
+            if (ActivityCompat.shouldShowRequestPermissionRationale(this, Manifest.permission.ACCESS_FINE_LOCATION)) {
+            } else {
+                ActivityCompat.requestPermissions(this, arrayOf(Manifest.permission.ACCESS_FINE_LOCATION), LOCATION_REQUEST_CODE)
+            }
+        }
     }
 
     override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<out String>, grantResults: IntArray) {
         super.onRequestPermissionsResult(requestCode, permissions, grantResults)
-        officeFeature.onRequestPermissionsResult(requestCode, grantResults)
+        when (requestCode) {
+            LOCATION_REQUEST_CODE -> {
+                if (grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
```

Comment:
Ja w 1. kolejności bym szedł w stronę m/w czegoś takiego: ` grantResults.elementAtOrElse(0, { PackageManager.PERMISSION_DENIED }) == PackageManager.PERMISSION_GRANTED`
ale czy to jest bardziej czytelne?